### PR TITLE
Use DI for all Antiforgery services

### PR DESCRIPTION
**Pull request by rynowak**
This change makes it possible to replace all of the various
IAntiforgery*** extensibility points via DI.

changes:
- Move functionality out of AntiforgeryWorker into Antiforgery
- Move services to DI (instead of constructed by Antiforgery)
- Cleanup how application/cookie-name is computed
- Merge IAntiforgeryTokenGenerator & IAntiforgeryTokenValidator
- Unseal classes
- Fix use of options in services
- Misc test cleanup

### CODE REVIEW

**Comment by rynowak** [comment]
/cc @Eilon 

After this change, this is probably *good enough* that we can start using it in MVC and then address various bugs/feature requests.

File: src/Microsoft.AspNet.Antiforgery/Antiforgery.cs [hunk 1]
```diff
@@ -20,27 +15,26 @@ namespace Microsoft.AspNet.Antiforgery
...
     {
-        private static readonly string _purpose = "Microsoft.AspNet.Antiforgery.AntiforgeryToken.v1";
-        private readonly AntiforgeryWorker _worker;
+        private readonly IHtmlEncoder _htmlEncoder;
+        private readonly AntiforgeryOptions _options;
+        private readonly IAntiforgeryTokenGenerator _tokenGenerator;
+        private readonly IAntiforgeryTokenSerializer _tokenSerializer;
+        private readonly IAntiforgeryTokenStore _tokenStore;
 
         public Antiforgery(
```
Comment (rynowak) [inline src/Microsoft.AspNet.Antiforgery/Antiforgery.cs]: This is 99% copy-paste from `AntiforgeryWorker`.

File: src/Microsoft.AspNet.Antiforgery/AntiforgeryTokenSerializer.cs [hunk 2]
```diff
@@ -9,14 +9,16 @@
...
+        private static readonly string Purpose = "Microsoft.AspNet.Antiforgery.AntiforgeryToken.v1";
+
         private readonly IDataProtector _cryptoSystem;
         private const byte TokenVersion = 0x01;
 
-        public AntiforgeryTokenSerializer([NotNull] IDataProtector cryptoSystem)
+        public AntiforgeryTokenSerializer([NotNull] IDataProtectionProvider provider)
         {
-            _cryptoSystem = cryptoSystem;
+            _cryptoSystem = provider.CreateProtector(Purpose);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Antiforgery/AntiforgeryTokenSerializer.cs]: Makes this DI friendly.

File: src/Microsoft.AspNet.Antiforgery/AntiforgeryTokenStore.cs [hunk 3]
```diff
@@ -6,74 +6,77 @@
...
-        private readonly AntiforgeryOptions _config;
-        private readonly IAntiforgeryTokenSerializer _serializer;
+        private readonly AntiforgeryOptions _options;
+        private readonly IAntiforgeryTokenSerializer _tokenSerializer;
 
-        public AntiforgeryTokenStore([NotNull] AntiforgeryOptions config,
-                                       [NotNull] IAntiforgeryTokenSerializer serializer)
+        public AntiforgeryTokenStore(
+            [NotNull] IOptions<AntiforgeryOptions> optionsAccessor,
+            [NotNull] IAntiforgeryTokenSerializer tokenSerializer)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Antiforgery/AntiforgeryTokenStore.cs]: Makes this DI friendly.

File: src/Microsoft.AspNet.Antiforgery/AntiforgeryTokenStore.cs [hunk 4]
```diff
@@ -6,74 +6,77 @@
...
+            _options = optionsAccessor.Options;
+            _tokenSerializer = tokenSerializer;
         }
 
         public AntiforgeryToken GetCookieToken(HttpContext httpContext)
         {
-            var contextAccessor =
-                httpContext.RequestServices.GetRequiredService<IAntiforgeryContextAccessor>();
+            var services = httpContext.RequestServices;
+            var contextAccessor = services.GetRequiredService<IAntiforgeryContextAccessor>();
```
Comment (rynowak) [inline src/Microsoft.AspNet.Antiforgery/AntiforgeryTokenStore.cs]: We should think about just using `.Items` here. This is exactly what it's for.

File: src/Microsoft.AspNet.Antiforgery/IAntiforgeryTokenValidator.cs [hunk 5]
```diff
@@ -1,23 +0,0 @@
-// Copyright (c) .NET Foundation. All rights reserved.
-// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
-
-using System.Security.Claims;
-using Microsoft.AspNet.Http;
-
-namespace Microsoft.AspNet.Antiforgery
-{
-    // Provides an abstraction around something that can validate anti-XSRF tokens
-    public interface IAntiforgeryTokenValidator
```
Comment (rynowak) [inline src/Microsoft.AspNet.Antiforgery/IAntiforgeryTokenValidator.cs]: Combined with `IAntiforgeryTokenGenerator`. If you can generate a token, you probably can validate it too, and the implementation code already reflected that.

File: src/Microsoft.AspNet.Antiforgery/Internal/AntiforgeryWorker.cs [hunk 6]
```diff
@@ -1,251 +0,0 @@
...
-using System.Diagnostics;
-using System.Security.Claims;
-using System.Threading.Tasks;
-using Microsoft.AspNet.Http;
-using Microsoft.Framework.Internal;
-using Microsoft.Framework.WebEncoders;
-
-namespace Microsoft.AspNet.Antiforgery.Internal
-{
-    public sealed class AntiforgeryWorker
```
Comment (rynowak) [inline src/Microsoft.AspNet.Antiforgery/Internal/AntiforgeryWorker.cs]: Moved into `Antiforgery`. We didn't need 2 facades.

File: src/Microsoft.AspNet.Antiforgery/ServiceCollectionExtensions.cs [hunk 7]
```diff
@@ -14,11 +15,15 @@ public static IServiceCollection AddAntiforgery([NotNull] this IServiceCollectio
             services.AddDataProtection();
             services.AddWebEncoders();
 
-            services.TryAdd(ServiceDescriptor.Singleton<IClaimUidExtractor, DefaultClaimUidExtractor>());
-            services.TryAdd(ServiceDescriptor.Singleton<Antiforgery, Antiforgery>());
-            services.TryAdd(ServiceDescriptor.Scoped<IAntiforgeryContextAccessor, AntiforgeryContextAccessor>());
-            services.TryAdd(
-                ServiceDescriptor.Singleton<IAntiforgeryAdditionalDataProvider, DefaultAntiforgeryAdditionalDataProvider>());
+            services.Add(ServiceDescriptor.Transient<IConfigureOptions<AntiforgeryOptions>, AntiforgeryOptionsSetup>());
```
Comment (rynowak) [inline src/Microsoft.AspNet.Antiforgery/ServiceCollectionExtensions.cs]: Need to use my new `TryAddEnumerable` once its available.

File: src/Microsoft.AspNet.Antiforgery/Antiforgery.cs [hunk 8]
```diff
@@ -20,27 +15,26 @@ namespace Microsoft.AspNet.Antiforgery
     /// Provides access to the anti-forgery system, which provides protection against
     /// Cross-site Request Forgery (XSRF, also called CSRF) attacks.
     /// </summary>
-    public sealed class Antiforgery
+    public class Antiforgery
```
Comment (rynowak) [inline src/Microsoft.AspNet.Antiforgery/Antiforgery.cs]: Should we come up with a better name for this?

**Comment by Eilon** [comment]
Looks :shipit: to me though I didn't get the comment about using `Items` or whatever.

**Pull request closed without merging**